schema.pgsql

```diff
@@ -16,14 +16,61 @@ CREATE INDEX files_expiry ON files(expiry);
 CREATE TABLE release_versions (
     project varchar(50) PRIMARY KEY,
     version varchar(25) NOT NULL,
+    prerelease_version varchar(25),
     updated TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW()
 );
 
 -- Abritrary version values at the (approx) time this was written; this don't really matter as
 -- they'll get updated within a first few seconds of initial startup.
-INSERT INTO release_versions (project, version, updated) VALUES ('oxen-io/session-desktop', 'v1.7.3', '2021-10-14Z');
-INSERT INTO release_versions (project, version, updated) VALUES ('oxen-io/session-android', '1.11.11', '2021-10-14Z');
-INSERT INTO release_versions (project, version, updated) VALUES ('oxen-io/session-ios', '1.11.17', '2021-10-14Z');
+INSERT INTO release_versions (project, version, prerelease_version, updated) VALUES ('oxen-io/session-desktop', 'v1.7.3', NULL, '2021-10-14Z');
+INSERT INTO release_versions (project, version, prerelease_version, updated) VALUES ('oxen-io/session-android', '1.11.11', NULL, '2021-10-14Z');
+INSERT INTO release_versions (project, version, prerelease_version, updated) VALUES ('oxen-io/session-ios', '1.11.17', NULL, '2021-10-14Z');
+
```

Comment:
There's some (undeclared) multi-column foreign keys here which we could normalize to clean up the structure a bit.  I played around with it a bit and came up with this:

```SQL
CREATE TABLE projects (id BIGSERIAL PRIMARY KEY, name varchar(50));

CREATE TABLE releases (
    id BIGSERIAL PRIMARY KEY,
    project BIGINT NOT NULL REFERENCES projects(id),
    prerelease BOOLEAN NOT NULL DEFAULT FALSE,
    version_code BIGINT NOT NULL,
    url varchar(255) NOT NULL,
    name varchar(255),
    notes text,
    UNIQUE(project, version_code)
);

CREATE TABLE release_assets (
    release BIGINT NOT NULL REFERENCES releases(id),
    name varchar(255) NOT NULL,
    url varchar(255) NOT NULL
);
CREATE INDEX ON release_assets(release);

CREATE VIEW versions AS
    SELECT
        releases.id as id,
        projects.name as proj_name,
        version_code,
        version_code / 1000000 || '.' || version_code % 1000000 / 1000 || '.' || version_code % 1000 AS version,
        prerelease,
        url,
        releases.name AS name,
        notes
    FROM releases JOIN projects ON releases.project = projects.id;

CREATE VIEW release_versions AS SELECT * FROM versions WHERE NOT prerelease;
CREATE VIEW prerelease_versions AS SELECT * FROM versions WHERE prerelease;
```

The `version_code` would be a numeric representation of the version, e.g. we'd insert 2.3.4 as `2003004` (and the VIEW, above, converts it back to the dotted form upon querying).

The whole thing then gets populated like this:

```SQL
insert into projects (name) values ('abc');
insert into releases (project, version_code, url, name, notes) values ((select id from projects where name = 'abc'), 2003004, 'http://example.com/download/me', 'Exciting new 2.3.4 update with 26% more hokeypokey', '- Added 20% more hokeypokey
- Added 10% even more hokeypokey
- Removed 5.5% hokeypokey');
insert into releases (project, prerelease, version_code, url, name, notes) values ((select id from projects where name = 'abc'), TRUE, 2003005, 'http://example.com/download/me/again', 'Less exciting 2.3.5 update with only 3% more hokeypokey', '- Experimental new 3% hokeypokey boost');
insert into releases (project, version_code, url, name, notes) values ((select id from projects where name = 'abc'), 2003006, 'http://example.com/download/me/again', 'Less exciting 2.3.6 update with only 3% more hokeypokey', '- Final release of new 3% hokeypokey boost');
```

And querying: looks like this:
```sql
=> select * from release_versions ORDER BY version_code DESC;
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+--------------------------------------------+
| id | proj_name | version_code | version | prerelease |                 url                  |                          name                           |                   notes                    |
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+--------------------------------------------+
|  3 | abc       |      2003006 | 2.3.6   | f          | http://example.com/download/me/again | Less exciting 2.3.6 update with only 3% more hokeypokey | - Final release of new 3% hokeypokey boost |
|  1 | abc       |      2003004 | 2.3.4   | f          | http://example.com/download/me       | Exciting new 2.3.4 update with 26% more hokeypokey      | - Added 20% more hokeypokey               +|
|    |           |              |         |            |                                      |                                                         | - Added 10% even more hokeypokey          +|
|    |           |              |         |            |                                      |                                                         | - Removed 5.5% hokeypokey                  |
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+--------------------------------------------+
(2 rows)

=> select * from prerelease_versions ORDER BY version_code DESC;
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+----------------------------------------+
| id | proj_name | version_code | version | prerelease |                 url                  |                          name                           |                 notes                  |
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+----------------------------------------+
|  2 | abc       |      2003005 | 2.3.5   | t          | http://example.com/download/me/again | Less exciting 2.3.5 update with only 3% more hokeypokey | - Experimental new 3% hokeypokey boost |
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+----------------------------------------+
(1 row)
```

Aside from better normalization in the database, this also simplifies things by no longer needing to worry about delete old release data.  Not deleting previous release data actually enables a nice capability in that if someone queries the latest version and passes their own version, we can include cumulative release notes for all the versions since the version they are currently on, by querying:
```sql
=> select * from release_versions WHERE version_code > 2003004 ORDER BY version_code DESC;
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+--------------------------------------------+
| id | proj_name | version_code | version | prerelease |                 url                  |                          name                           |                   notes                    |
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+--------------------------------------------+
|  4 | abc       |      3000000 | 3.0.0   | f          | http://example.com/3.0.0             | 3.0.0 release                                           | - New release to make the number bigger    |
|  3 | abc       |      2003006 | 2.3.6   | f          | http://example.com/download/me/again | Less exciting 2.3.6 update with only 3% more hokeypokey | - Final release of new 3% hokeypokey boost |
+----+-----------+--------------+---------+------------+--------------------------------------+---------------------------------------------------------+--------------------------------------------+
(2 rows)
```

Reply:
Yea nice, I'll make these changes now - since I don't have a lot of Python experience and hadn't touched the File Server code before I was trying to minimise the changes I made 😁